monai/engines/workflow.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+import torch
+from ignite.engine import Engine, State, Events
+from .utils import default_prepare_batch
+
+
+class Workflow(ABC):
+    """Workflow defines the core work process totally based on Ignite engine.
+    All trainer, validator and evaluator share this same workflow as base class,
+    because they all can be treated as same Ignite engine loops.
+    And it initializes all the sharable data in Ignite engine.state.
+    Attach additional processing logics to Ignite engine based on Event-Handler mechanism.
+
+    Args:
+        device (torch.device): an object representing the device on which to run.
+        max_epochs (int): the total epoch number for engine to run, validator and evaluator have only 1 epoch.
+        amp (bool): whether to enable auto-mixed-precision training.
+        data_loader (torch.DataLoader): Ignite engine use data_loader to run, must be torch.DataLoader.
+        prepare_batch (Callable): function to parse image and label for current iteration.
+        key_metric (ignite.metric): compute metric when every iteration completed, and save average
+            value to engine.state.metrics when epoch completed. also use key_metric to select and
+            save checkpoint into files.
+        additional_metrics (list): more ignite metrics that also attach to Ignite Engine.
+        handlers (list): every handler is a set of Ignite Event-Handlers, like:
+            CheckpointHandler, StatsHandler, TimerHandler, etc.
+
+    """
+
+    def __init__(
```

Comment:
It would be nice to have an optional callable parameter which is used as the step function for the `Engine`, if not given `self._iteration` is used as the default.

Reply:
added step_function in the latest patch.
Thanks.